pvlib/modelchain.py

```diff
@@ -810,32 +810,101 @@ def run_model(self, times=None, weather=None):
         Parameters
         ----------
         times : None or DatetimeIndex, default None
-            Times at which to evaluate the model. Can be None if
-            attribute `times` is already set.
+            Times at which to evaluate the model. Can be None if attribute
+            `times` is already set.
         weather : None or DataFrame, default None
-            If None, assumes air temperature is 20 C, wind speed is 0
-            m/s and irradiation calculated from clear sky data. Column
-            names must be 'wind_speed', 'temp_air', 'dni', 'ghi', 'dhi'.
-            Do not pass incomplete irradiation data. Use method
-            :py:meth:`~pvlib.modelchain.ModelChain.complete_irradiance`
-            instead.
+            If None, assumes air temperature is 20 C, wind speed is 0 m/s and
+            irradiation calculated from clear sky data.
+            For conventional weather data, column names must be 'wind_speed',
+            'temp_air', 'dni', 'ghi', 'dhi'. Do not pass incomplete irradiation
+            data, instead use method
+            :py:meth:`~pvlib.modelchain.ModelChain.complete_irradiance`.
+            For complete PV reference device data, column names must be 'F' and
+            'H', for effective irradiance ratio and effective temperature
+            ratio, respectively. See :py:func:`~pvlib.pvsystem.sdm_campanelli`.
+            If either 'F' or 'H' is missing, then convential weather data are
+            also required to compute the missing 'F' or 'H'.
 
         Returns
         -------
         self
 
-        Assigns attributes: times, solar_position, airmass, irradiance,
-        total_irrad, effective_irradiance, weather, temps, aoi,
-        aoi_modifier, spectral_modifier, dc, ac, losses.
+        For conventional weather data, assigns attributes: times,
+        solar_position, airmass, irradiance, total_irrad, effective_irradiance,
+        weather, temps, aoi, aoi_modifier, spectral_modifier, dc, ac, losses.
+        For complete reference device data, only assigns attributes: times, dc,
+        ac, losses.
         """
 
-        self.prepare_inputs(times, weather)
-        self.aoi_model()
-        self.spectral_model()
-        self.effective_irradiance_model()
-        self.temp_model()
+        if weather is not None and 'F' in weather:
+            # Use PV reference device weather information for F
+            self.F = weather['F']
+        else:
+            # Clear any stale value
+            if hasattr(self, 'F'):
+                print('F cleared')
+                del self.F
```

Comment:
maybe this is debugging code, but we don't do this for any other modelchain attributes and I think we should avoid adding this layer of complexity in general.

Reply:
Yep. I forgot to remove the print statements, but in general these do need to be cleared to prevent subtle bugs that the user could run into when re-running a model with changing weather types, because I think it is generally valuable to "record" the `F` and `H` values used in the `ModelChain` object, which sometimes come from the "non-trivial" SAPM model calculations. (This is not too different than other "recorded" outputs, such as effective_irradiance.)